synapse/federation/federation_client.py

```diff
@@ -744,15 +755,10 @@ async def _do_send_join(self, destination: str, pdu: EventBase) -> JsonDict:
                 content=pdu.get_pdu_json(time_now),
             )
         except HttpResponseException as e:
-            if e.code in [400, 404]:
-                err = e.to_synapse_error()
-
-                # If we receive an error response that isn't a generic error, or an
-                # unrecognised endpoint error, we  assume that the remote understands
-                # the v2 invite API and this is a legitimate error.
-                if err.errcode not in [Codes.UNKNOWN, Codes.UNRECOGNIZED]:
-                    raise
-            else:
+            # If an error is received that is due to an unrecognised endpoint,
+            # fallback to the v1 endpoint. Otherwise consider it a legitmate error
+            # and raise.
+            if not self._is_unknown_endpoint(e, e.to_synapse_error()):
```

Comment:
it's tempting to move the `to_synapse_error` call into `_is_unknown_endpoint` even though that causes duplication in `_try_destination_list`. I feel like that makes `_is_unknown_endpoint` clearer.

Reply:
I also debated whether we could make it optional and generate it inside of `_is_unknown_endpoint` if it isn't provided?